padertorch/data/segment.py

```diff
@@ -0,0 +1,413 @@
+from paderbox.utils.nested import flatten, deflatten
+from padertorch.utils import to_list
+
+possible_anchor_modes = [
+    'left',
+    'right',
+    'center',
+    'centered_cutout',
+    'random',
+    'random_max_segments',
+]
+
+
+class Segmenter:
+    """
+    This segmenting returns a list of segmented examples that can be unbatched.
+    Everything that is not listed in `keys` is simply copied from the
+    input example to the output examples.
+    The keys `segment_start` and `segment_stop` are added for each output
+    dictionary.
+
+    If an utterance is shorter than `length`, a
+    `lazy_dataset.FilterException` is raised.
+
+    Examples (For more examples see tests/test_data/test_segmenter:
+
+        >>> segmenter = Segmenter(length=32000, include_keys=('x', 'y'),\
+                                   shift=16000)
+        >>> ex = {'x': np.arange(65000), 'y': np.arange(65000),\
+                    'num_samples': 65000, 'gender': 'm'}
+        >>> segmented = segmenter(ex)
+        >>> type(segmented)
+        <class 'list'>
+        >>> for entry in segmented:
+        ...     print(entry['x'][[0, -1]])
+        [    0 31999]
+        [16000 47999]
+        [32000 63999]
+
+        segmenting can be disabled by setting `length=-1` or `keys=None`
+        >>> Segmenter(length=-1, include_keys=('x', 'y'))(ex)[0] == ex
+        True
+
+        Check the corner cases.
+        >>> ex = {'x': np.arange(64000), 'y': np.arange(64000)}
+        >>> for entry in segmenter(ex):
+        ...     print(entry['x'][[0, -1]])
+        [    0 31999]
+        [16000 47999]
+        [32000 63999]
+        >>> ex = {'x': np.arange(63999), 'y': np.arange(63999)}
+        >>> for entry in segmenter(ex):
+        ...     print(entry['x'][[0, -1]])
+        [    0 31999]
+        [16000 47999]
+
+    Args:
+        length: The length of the segments in samples. If set to `-1`,
+            the original example is returned in a list of length one.
+        shift: shift between segments, defaults to length
+        include_keys: The keys in the passed example dict to segment. They all
+            must have the same size along their specified `axis`, if keys is
+            None the segmentation is applied to all `numpy.arrays`.
+            If a key points to a dictionary the segmentation is applied to all
+            values of this dictionary
+        exclude_keys: This option allows to specify specific keys not to
+            segment. This might be usefull if `include_keys` is `None` or
+            one of the included keys points to a dictionary.
+        copy_keys: If `None` all not values which are not segmented are
+            copied. Otherwise only the specified keys are added to the new
+            dictionary with the segmented signals.
+        axis: axis over which to segment
+        anchor_mode: anchor mode used in `get_anchor` to calculate the anchor
+            from which the segmentation boundaries are calculated.
```

Comment:
I thought about that, but I could not think of a usecase.

Reply:
I currently can't think of a use-case either, maybe someone wants to remove a short silence in the beginning? Anyway, I just thought it would be nice to have the same interface as `segment` which allows setting an integer.

Reply:
Sure, I can add that it is not much of a change